File: src/tests/3 LCS/main.c
#include <stdio.h>
#include <string.h>
#include <sys/time.h>

char X[2047];             // input
unsigned int n;           // length of X
char Y[2047];             // input
unsigned int m;           // length of Y
unsigned T;               // time limit (seconds)
struct timeval stopTime;  // time when we should stop
unsigned long iterations; // amount of finished cycles of computing

unsigned int res[2048][2048]; // array of subsquences' max lengths

void start();
void reinit();
void calc();
void outputLCS();

int main (int argc, char ** argv) {
  if (argc != 2) {
    printf("Time limit parameter required:\n ./main.bin <time>\n");
  } else {
    iterations = 0;
    sscanf(argv[1], "%d", &T);
    printf("Calculating LCS of two strings X and Y many times during %u seconds\n", T);
    printf("Input X: ");

    fgets(X, 2047, stdin);
    n = (unsigned int)strlen(X);
    X[--n] = 0; // trim new-line
    printf("Input Y: ");
    fgets(Y, 2047, stdin);
    m = (unsigned int)strlen(Y);
    Y[--m] = 0; // trim new-line
    printf("\n");
    start();
  }
  return 0;
}

void start() {
  struct timeval now;
  gettimeofday(&stopTime, NULL);
  stopTime.tv_sec += T; // NOW + T seconds
  gettimeofday(&now, NULL);
  while (now.tv_sec <  stopTime.tv_sec ||
        (now.tv_sec == stopTime.tv_sec && now.tv_usec < stopTime.tv_usec)) {
    reinit();
    calc();
    iterations++;
    gettimeofday(&now, NULL);
  }
  iterations--; // last iteration started before time was up but finished after time was up
  // required '>' mark for automatic benchmark
  printf("> %lu iterations\n", iterations);
  printf("Resulting LCS(X, Y) = %u\n", res[n][m]);
  outputLCS();
}

void reinit() {
  // already allocated (limited by 2048 x 2048)
  // no cleanup needed
}

void calc() {
  for (unsigned int i = 0; i <= n; i++) {
    res [i][0] = 0;
  }
  for (unsigned int i = 0; i <= m; i++) {
    res [0][i] = 0;
  }
  for (unsigned int i = 1; i <= n; i++) {
    for (unsigned int j = 1; j <= m; j++) {
      if (X[i-1] == Y[j-1] ) {
        res [i][j] = 1 + res [i-1][j-1];
      } else {
        if (res[i-1][j] > res[i][j-1]) {
          res [i][j] = res[i-1][j];
	} else {
          res [i][j] = res[i][j-1];
	}
      }
    }
  }
}

void outputLCS() {
  unsigned int i = n, j = m;
  char substrX[2047];
  char substrY[2047];
  for (unsigned int a = 0; a < n; a++) {
    substrX[a] = '_';
  }
  substrX[n] = 0;
  for (unsigned int a = 0; a < m; a++) {
    substrY[a] = '_';
  }
  substrY[n] = 0;

  while ( i > 0 && j > 0 ) {
    if (X[i-1] == Y[j-1]) {
      i--; j--;
      substrX[i] = X[i];
      substrY[j] = Y[j];
    } else {
      if ( res[i-1][j] > res[i][j-1] ) {
        i--;
      } else {
        j--;
      }
    }
  }

  printf("\n%s   %s\n", X, Y);
  printf("%s = %s\n", substrX, substrY);
}

File: src/tests/3 LCS/main.php
<?php
error_reporting(E_ALL);
$X = "";         // input
$n = 0;          // length of X
$Y = "";         // input
$m = 0;          // length of Y
$T = 0;          // time limit (seconds)
$stopTime = 0;   // time when we should stop
$iterations = 0; // amount of finished cycles of computing

$res = NULL;     // array of subsquences' max lengths

if (count($argv) != 2) {
  print("Time limit parameter required:\n php main.php <time>\n");
} else {
  $iterations = 0;
  $T = $argv[1];
  print("Calculating LCS of two strings X and Y many times during $T seconds\n");
  print("Input X:\n");
  
  $X = trim(fgets(STDIN, 2047));
  $n = strlen($X);
  print("Input Y:\n");

  $Y = trim(fgets(STDIN, 2047));
  $m = strlen($Y);
  start();
}

function start() {
  global $stopTime, $iterations, $T, $res, $n, $m;
  $stopTime = microtime(true) + $T; // NOW + T seconds
  while ( microtime(true) < $stopTime) {
    reinit();
    calc();
    $iterations++;
  }
  $iterations--; // last iteration started before time was up but finished after time was up
  // required '>' mark for automatic benchmark
  print("> $iterations iterations\n");
  print("Resulting LCS(X, Y) = {$res[$n][$m]}\n");
  outputLCS();
}

function reinit () {
  global $res, $n, $m;
  if ($res === NULL) {
    // create 2-dimensional array (n+1) x (m+1)
    $res = array(array());
    for ($i = 1; $i <= $n; $i++) {
      $res[$i] = array($m=>0);
    }
  }
  // no cleanup needed
}

function calc() {
  global $res, $n, $m, $X, $Y;
  for ($i = 0; $i <= $n; $i++) {
    $res[$i][0] = 0;
  }
  for ($i = 0; $i <= $m; $i++) {
    $res[0][$i] = 0;
  }
  for ($i = 1; $i <= $n; $i++) {
    for ($j = 1; $j <= $m; $j++) {
      if ($X[$i - 1] === $Y[$j - 1]) {
        $res[$i][$j] = 1 + $res[$i-1][$j-1];
      } else {
        $res[$i][$j] = max($res[$i-1][$j], $res[$i][$j-1]);
      }
    }
  }
}

function outputLCS() {
  global $n, $m, $res, $X, $Y;
  $i = $n;
  $j = $m;
  $substrX = preg_replace("/./", "_", $X);
  $substrY = preg_replace("/./", "_", $Y);

  while ( $i > 0 && $j > 0) {
    if ($X[$i-1] === $Y[$j-1]) {
      $i--;
      $j--;
      $substrX[$i] = $X[$i];
      $substrY[$j] = $Y[$j];
    } else {
      if ($res[$i-1][$j] > $res[$i][$j-1]) {
        $i--;
      } else {
        $j--;
      }
    }
  }

  print("\n$X   $Y\n");
  print("$substrX = $substrY\n");
}
